### ISSUE TITLE
Feature Request: Sum Aggregator for the `map` type

### ISSUE THREAD

**Issue by jeremiah-corrado** [chapel-lang/chapel#24377]
With the existing `map` interface, parallel map reductions can be somewhat difficult to express. Consider the following serial code that reduces an array of `(string, int)` pairs into a map by summing the integer values associated with each unique string:

```chapel
use Map;

const inputData: [1..1000] (string, int) = getSomeInputData();
var m = new map(string, int);

for (k, v) in inputData do m[k] += v;
```

Today, parallelizing the above code might look like the following. In this program, a parallel safe map is used. This exposes the `update` interface, where a callable-record/FCF can be used to atomically update a value in the map:
```chapel
use Map;

const inputData: [1..1000] (string, int) = getSomeInputData();
var m = new map(string, int, parSafe=true);

forall (k, v) in inputData {
  if m.contains(k)
    then m.update(k, new addOp(v));
    else if !m.add(k, v) then m.update(k, new addOp(v));
}

record addOp {
  var val: int;
  proc this(k: string, ref v: int) { v += val; return none; }
}
```

This code is quite a bit more involved than the serial version. Part of the problem is that `update` can only operate on keys that are already in the map, so checking whether the key exists before calling `update` is required.

Additionally, since a parallel safe map must be used here, each method call on the map involves taking and releasing the map's global lock (potentially up to 3 times per iteration). This form of synchronization won't scale well, so it would be up to the user to manually come up with a more scalable parallelization strategy.

An alternative might be to support a reducing aggregator supplied by the map type itself. This would allow one to write something like the following:
```chapel
use Map;

const inputData: [1..1000] (string, int) = getSomeInputData();
var m = new map(string, int);

forall (k, v) in inputData with (var agg = m.sumAggregator()) do agg.addReduce(k, v);
```

From a user perspective, this interface clearly expresses the operation being computed and follows pretty directly from the serial code. From an implementation perspective, it allows the map type to handle performance and scalability, potentially leveraging details about the underlying implementation that a user wouldn't have access to.

Related: https://github.com/chapel-lang/chapel/issues/9848

**Comment by mppf** [comment]
@lydia-duncan - I recall that you had some prototypes of map aggregators and we had some discussion about what would be reasonable to add for distributed computation. Could you help us to recall this information?

My main concern about the approach here is that `parSafe` maps are unstable

> 'map.parSafe' is unstable and is expected to be replaced by a separate map type in the future

That leads me to wonder if we would want this new functionality only on the separate type. Of course it's interesting to consider that perhaps the aggregator pattern can work even with a non-parsafe map, because it's the aggregator that is doing the locking/isolation. That leads me to wonder if such aggregators should be separate from the `map` type.

A secondary concern is -- what if the user does not want to add the values when combining maps and wants to do something else? How can the interface make sense for that?

**Comment by jeremiah-corrado** [comment]
> Of course it's interesting to consider that perhaps the aggregator pattern can work even with a non-parsafe map, because it's the aggregator that is doing the locking/isolation.

I think that approach could work (with the caveat that the map can't be manipulated by these aggregators and some other methods concurrently).

> A secondary concern is -- what if the user does not want to add the values when combining maps and wants to do something else? How can the interface make sense for that?

Other than having multiple aggregators for different operations, I think the user could use a different type with their own definition of `+=`:

```chapel
var m = new map(string, rec);

forall (k, v) in inputData with (var agg = m.sumAggregator()) do agg.addReduce(k, v);

record rec { val x: int; }
operator rec.+=(ref left: rec, right: int) do left.x *= right; // do multiplication instead
```

**Comment by lydia-duncan** [comment]
Yeah, the [v2 implementation](https://github.com/chapel-lang/chapel/blob/main/test/library/draft/DistributedMap/v2/DistributedMap.chpl) looked at interplay with aggregators, where [updateAggregator](https://github.com/chapel-lang/chapel/blob/main/test/library/draft/DistributedMap/v2/DistributedMap.chpl#L717) would take in a first class procedure to use for updating the contents of the map and return a general aggregator that would use that FCP on the specified key's value (see [this test](https://github.com/chapel-lang/chapel/blob/main/test/library/draft/DistributedMap/v2/use-distributed-map.chpl) for a basic mock up of using it).  This had the benefit of being agnostic towards the update operation in question.

I think it would make more sense to fill out the ParallelMap/DistributedMap implementation than put something on the serial map using parSafe since it's unstable and also since the current synchronization style is not ideal.  Might as well make the better synchronization style in the separate type like we were planning than force ourselves to use the part that's going away.

**Comment by lydia-duncan** [comment]
The interface bones should be pretty guided by the serial map as it stands (after all, that was the point of stabilizing it this way), so there shouldn't need to be too much discussion beyond what to do about the interface using the aggregator, where I think the mockup I pointed to was a reasonable starting point

**Comment by mppf** [comment]
Note also that we have modules/packages/ConcurrentMap to try out and also a distributed version of the same at https://github.com/dgarvit/distributed-interlocked-hash-table . We should at least compare the performance of these approaches to the various aggregator implementations.
